contrib/config/source/kv_store_xds_delegate.cc

```diff
@@ -0,0 +1,146 @@
+#include "contrib/config/source/kv_store_xds_delegate.h"
+
+#include "envoy/registry/registry.h"
+#include "envoy/service/discovery/v3/discovery.pb.h"
+
+#include "source/common/common/logger.h"
+#include "source/common/config/utility.h"
+#include "source/common/protobuf/utility.h"
+
+#include "absl/strings/match.h"
+#include "absl/strings/str_cat.h"
+#include "absl/types/optional.h"
+#include "contrib/envoy/extensions/config/v3alpha/kv_store_xds_delegate_config.pb.h"
+#include "contrib/envoy/extensions/config/v3alpha/kv_store_xds_delegate_config.pb.validate.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Config {
+namespace {
+
+using ::Envoy::Config::XdsSourceId;
+using ::envoy::extensions::config::v3alpha::KeyValueStoreXdsDelegateConfig;
+
+// Constructs the key for the resource, to be used in the KeyValueStore.
+std::string constructKey(const XdsSourceId& source_id, const std::string& resource_name) {
+  static constexpr char DELIMITER[] = "+";
+  return absl::StrCat(source_id.toKey(), DELIMITER, resource_name);
+}
+
+} // namespace
+
+XdsKeyValueStoreStats KeyValueStoreXdsDelegate::generateStats(Stats::Scope& scope) {
+  return {ALL_XDS_KV_STORE_STATS(POOL_COUNTER(scope))};
+}
+
+KeyValueStoreXdsDelegate::KeyValueStoreXdsDelegate(KeyValueStorePtr&& xds_config_store,
+                                                   Stats::Scope& root_scope)
+    : xds_config_store_(std::move(xds_config_store)),
+      scope_(root_scope.createScope("xds.kv_store.")), stats_(generateStats(*scope_)) {}
+
+std::vector<envoy::service::discovery::v3::Resource>
+KeyValueStoreXdsDelegate::getResources(const XdsSourceId& source_id,
+                                       const std::vector<std::string>& resource_names) const {
+  std::vector<envoy::service::discovery::v3::Resource> resources;
+  if (resource_names.empty() || (resource_names.size() == 1 && resource_names[0] == "*")) {
+    // Empty names or one entry with "*" means wildcard.
+    resources = getAllResources(source_id);
+  } else {
+    for (const std::string& resource_name : resource_names) {
+      const std::string resource_key = constructKey(source_id, resource_name);
+      if (const auto existing_resource = xds_config_store_->get(resource_key)) {
+        envoy::service::discovery::v3::Resource r;
+        if (!r.ParseFromString(std::string(*existing_resource))) {
```

Comment:
@adisuissa I added a check here if the resource fails to parse, including maintaining stats about it and removing the resource from the KV store if it's no longer parseable.

Reply:
I pushed a commit that shows via a test that we handle invalid protos (e.g. proto enum value no longer valid) appropriately.